client/device/cloudftd/upgrade.go

```diff
@@ -46,11 +49,52 @@ func (f *ftdUpgradeService) Upgrade(uid string, softwareVersion string) (*FtdDev
 	if err != nil {
 		return nil, err
 	}
-	err = f.validateFtdVersion(ftdDevice, softwareVersion)
+	err = f.validateFtdVersion(ftdDevice, softwareVersionStr)
 	if err != nil {
 		return nil, err
 	}
 
+	tflog.Debug(f.Ctx, "Triggering the FTD device upgrade.")
+	return f.doUpgrade(ftdDevice, softwareVersionStr)
+}
+
+func (f *ftdUpgradeService) doUpgrade(ftdDevice *FtdDevice, softwareVersionStr string) (*FtdDevice, error) {
+	versionOnDevice, err := ftd.NewVersion(ftdDevice.SoftwareVersion)
+	softwareVersion, err := ftd.NewVersion(softwareVersionStr)
+	if versionOnDevice.Equal(softwareVersion) {
+		return ftdDevice, nil
+	}
+	if err != nil {
+		f.Client.Logger.Printf("error parsing software version %s to upgrade to\n", softwareVersionStr)
+		return nil, err
+	}
+	upgradePackage, err := f.getUpgradePackage(ftdDevice, softwareVersion)
```

Comment:
See comments in validateUpgradePackage and getUpgradePackage -- you can completely remove this.